core/src/sumeragi/main_loop.rs

```diff
@@ -216,101 +224,206 @@ impl<F: FaultInjection> SumeragiWithFault<F> {
+                            // genesis round we blindly take on the network
+                            // topology described in the provided genesis
+                            // block.
+                            let block = {
+                                let span = span!(
+                                    Level::TRACE,
+                                    "Genesis Round Peer is revalidating the block."
+                                );
+                                let _enter = span.enter();
+                                match block.revalidate(
+                                    &self.transaction_validator,
+                                    &state.wsv,
+                                    state.latest_block_hash,
+                                    state.latest_block_height,
+                                ) {
+                                    Ok(block) => block,
+                                    Err(error) => {
+                                        error!(?error);
+                                        continue;
+                                    }
+                                }
+                            };
+                            // Omit signature verification during genesis round
+                            block.commit_unchecked().into()
+                        }
+                        Message::BlockSyncUpdate(BlockSyncUpdate { block }) => {
+                            // Omit signature verification during genesis round
+                            match block.revalidate_hashes() {
+                                Ok(block) => block,
+                                Err((_, error)) => {
+                                    error!(?error);
+                                    continue;
+                                }
+                            }
+                        }
+                        msg => {
+                            trace!(?msg, "Not handling the message, waiting for genesis...");
+                            continue;
+                        }
+                    };
+
+                    if block.header().is_genesis() {
+                        if let Some(topology) = block.header().genesis_topology.clone().take() {
+                            state.current_topology = topology;
+                            info!("Using genesis topology");
+                        }
+
+                        commit_block(self, state, block);
+                        return Err(EarlyReturn::GenesisBlockReceivedAndCommitted);
+                    }
+                    debug!("Received a block that was not genesis.");
+                }
+                Err(mpsc::TryRecvError::Disconnected) => return Err(EarlyReturn::Disconnected),
+                _ => (),
+            }
+        }
+    }
 }
 
-#[allow(clippy::expect_used, clippy::ptr_arg)]
-fn gossip_transactions<F>(
+fn gossip_transactions<F: FaultInjection>(
     sumeragi: &SumeragiWithFault<F>,
```

Comment:
What do you think of making this method of `sumeragi`?

Reply:
done, most of these functions can be made methods of `sumeragi`. This would require another round of refactor